src/dmd/dtemplate.d

```diff
@@ -4324,6 +4324,14 @@ MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplateParameters* param
                 }
             }
 
+            bool isTopRef(Type t)
+            {
+                auto tb = t.baseElemOf();
+                return tb.ty == Tclass ||
+                       tb.ty == Taarray ||
+                       tb.ty == Tstruct && tb.hasPointers();
```

Comment:
Do we need fail tests for these types?

Reply:
The failure conditions are covered using the `compiles` traits (e.g. `static assert(!__traits(compiles, bar!U(u)), U.stringof);`), in the runnable test.  Is that not sufficient?

Reply:
On second look, ok, there are three types.

- `Object` covers `tb.ty == Tclass`
- `Y` covers `tb.ty == Tstruct && tb.hasPointers()`
- `Y[2]` covers `t.baseElemOf()`

So only missing a check for AAs, adding `int[int]` should be enough by the looks.

Reply:
Done.  Thank you!